feat: support AnalyticDB vector store

lpdink opened the pull request:
# Description

- AnalyticDB offers enterprise-level vector database services tailored for generative AI applications. For further information, please see [here](https://www.alibabacloud.com/en/product/hybriddb-postgresql?_p_lc=1)
- This PR utilizes `alibabacloud_gpdb20160503` and `alibabacloud_tea_openapi` for database connectivity. It leverages a pure HTTP client architecture, hence a Docker image is not provided. 
- Implemented and successfully executed AnalyticDBTest using certain AK/SK.
- Due to local issues with Poetry, this PR does not directly add new dependencies to pyproject.toml. Could a maintainer kindly add the following dependencies?
```sh
poetry add alibabacloud_gpdb20160503="~3.8.0"
poetry add alibabacloud_tea_openapi="~0.3.9"
```

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Added the AnalyticdbVectorTest integration test which successfully passed using certain service credentials (AK/SK).

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods
- [ ] `optional` I have made corresponding changes to the documentation 
- [x] `optional` I have added tests that prove my fix is effective or that my feature works
- [x] `optional` New and existing unit tests pass locally with my changes

bowenliang123 commented on api/configs/middleware/__init__.py:
```diff
@@ -171,6 +172,7 @@ class MiddlewareConfig(
     # configs of vdb and vdb providers
     VectorStoreConfig,
     ChromaConfig,
+    AnalyticdbConfig,
```
```suggestion
    AnalyticdbConfig,
    ChromaConfig,
```

bowenliang123 commented on api/configs/middleware/vdb/analyticdb_config.py:
```diff
@@ -0,0 +1,42 @@
+    """
+
+    ANALYTICDB_KEY_ID : Optional[str] = Field(
+        description='Analyticdb key_id',
```
Hi, try to put the text in a more human-readable way for explanation and hints, rather than just copying the exact id of the field. Be nice, be informative.

lpdink replied:
Done. Thank you.

bowenliang123 replied:
Thanks!

bowenliang123 commented on api/configs/middleware/vdb/analyticdb_config.py:
```diff
@@ -0,0 +1,42 @@
+        default=None,
+    )
+    ANALYTICDB_REGION_ID : Optional[str] = Field(
+        description='Analyticdb region_id',
```
ditto.

bowenliang123 commented on api/configs/middleware/vdb/analyticdb_config.py:
```diff
@@ -0,0 +1,42 @@
+        default=None,
+    )
+    ANALYTICDB_INSTANCE_ID : Optional[str] = Field(
+        description='Analyticdb instance_id',
```
ditto.

bowenliang123 commented on api/configs/middleware/vdb/analyticdb_config.py:
```diff
@@ -0,0 +1,42 @@
+        default=None,
+    )
+    ANALYTICDB_NAMESPACE_PASSWORD : Optional[str] = Field(
+        description='Analyticdb namespace_password',
```
ditto.

bowenliang123 commented on api/configs/middleware/vdb/analyticdb_config.py:
```diff
@@ -0,0 +1,42 @@
+
+class AnalyticdbConfig(BaseModel):
+    """
+    Analyticdb configs
```
Any possible official ADB docs to refer for the definitions or format of the concept and values?

bowenliang123 commented:
As for the dependencies, add the required packages to the `pyproject.toml`, and then use `dev/sync-poetry` to update the poetry lockfile.

And I'm new to ADB concepts, any reference or ADB docs for  `alibabacloud_gpdb20160503` and `alibabacloud_tea_openapi `. The package name sound abnormal to me especially the exact date inside the name which may indicate an version or compatibility issue? Why we would have it?

bowenliang123 commented on api/poetry.lock:
```diff
@@ -8238,20 +8441,19 @@ files = [
 
 [[package]]
 name = "urllib3"
-version = "2.0.7"
+version = "1.26.19"
```
It's a surprising downgrade in `urllib3` caused and required by `alibabacloud-tea` .

lpdink replied:
Indeed, it's quite strange. The `alibabacloud-tea` package requires a version of `urllib3` that should be `urllib3<3,>=1.21.1`. I don't quite understand why Poetry downgraded the version to 1.26.19.

bowenliang123 replied:
![image](https://github.com/langgenius/dify/assets/1935105/a261f6e7-45cd-4ae6-8c60-66baa9ec7055)
From the poetry lockfile, `alibabacloud-tea` does enforce to require urllib3 1.x.

bowenliang123 replied:
How about use the latest minor version first, like `alibabacloud_tea_openapi="~0.3.10"`

lpdink replied:
I have identified the source of the error, which is due to an inappropriate `egg-info/requires.txt` from the `alibabacloud-tea` dependency that `alibabacloud_tea_openapi` relies on. The `setup.py` for the `alibabacloud-tea` library checks the OpenSSL version and adds the requirement for `urllib3<2.0.0` if OpenSSL is missing or the version is too low. You can see this [here](https://github.com/aliyun/tea-python/blob/8073e26f8523935e068eb25e5c16e7236ca92608/setup.py#L41).   
Unfortunately, it seems that the machine used for packaging and uploading to PyPI did not have OpenSSL installed, causing this requirement to be unconditionally written into the `egg-info/requires.txt`, which was then downloaded and used by Poetry for dependency analysis.   
I have reached out to the developers of `alibabacloud-tea`, and they have indicated that the issue may be corrected in the next couple of days.   
Please check back here later. Thank you for your patience.

bowenliang123 replied:
I'm okay with it, if this downgrade is necessary. It would be just nice to be clarified in description or somewhere else to show the major changes of this PR.

bowenliang123 replied:
Btw, what does the word `tea` inside the package name stand for ? I am green to Aliyun or ADB's service.

lpdink replied:
err...I'm not sure, but it seems to stand for "Tea API Gateway", a tool to work with alibabacloud api, defined in [alibabacloud-apigateway-core-sdk](https://github.com/aliyun/alibabacloud-apigateway-core-sdk)

bowenliang123 replied:
Hard to understand , but still good to know then. Thanks, haha.

lpdink commented:
@bowenliang123 

> As for the dependencies, add the required packages to the `pyproject.toml`, and then use `dev/sync-poetry` to update the poetry lockfile.
> 
> And I'm new to ADB concepts, any reference or ADB docs for `alibabacloud_gpdb20160503` and `alibabacloud_tea_openapi `. The package name sound abnormal to me especially the exact date inside the name which may indicate an version or compatibility issue? Why we would have it?

Done, Thank you for your suggestion. I added the specific description and updated `pyproject.toml` . As for the naming of `alibabacloud_gpdb20160503 `, this is named according to the implementation time of the first API, but the module itself is newest and continuously maintained. The quick start of AnalyticDB also added to config, [here](https://www.alibabacloud.com/help/en/analyticdb-for-postgresql/getting-started/create-an-instance-instances-with-vector-engine-optimization-enabled)

bowenliang123 commented:
cc @JohnJyong

bowenliang123 commented:
https://help.aliyun.com/zh/analyticdb-for-postgresql/user-guide/sdk-python-3

lpdink commented:
@bowenliang123 Hey, The alibabacloud-tea has released version 0.3.7, and I've updated the poetry.lock. Now, urllib3 will no longer be downgraded. Could you please approve with this PR?

lpdink commented:
@laipz8200 Hey, could you please review this PR?

lpdink commented:
@JohnJyong Hey, I'm really sorry to bother you, but could you please kindly review this PR?

bowenliang123 commented:
LGTM overall. And how about resolving the conflicts first, considering many changes in configs and docker file merged recently.

lpdink committed:
feat: support AnalyticDB vector store

lpdink commented:
@bowenliang123 I'm really sorry to bother you, but could you please remind the other reviewers to approve this PR?

bowenliang123 commented:
Hi @lpdink , this PR is good for me. Feel free to ask Dify's member to review this PR, as I'm not one of them although I had some efforts in vdb features implementation and tests. cc @takatost @JohnJyong

Gaocarri commented:
@lpdink sorry to bother u, have you tried migrating pg to adb? I encountered some problems

lpdink commented:
> @lpdink sorry to bother u, have you tried migrating pg to adb? I encountered some problems

@Gaocarri I'm glad to help. Could you provide more details about the problem ?

Gaocarri commented:
> > @lpdink sorry to bother u, have you tried migrating pg to adb? I encountered some problems
> 
> @Gaocarri I'm glad to help. Could you provide more details about the problem ?

thanks your help!

i create a discussion for my problem: https://github.com/langgenius/dify/discussions/6720

i changed the docker/.env
```
DB_USERNAME=xxx
DB_PASSWORD=xxx
DB_HOST=xx
DB_PORT=xx
DB_DATABASE=xx
```

i can use psql to connect adb in container, but some troubles。
